lib/linked_list.rb

```diff
@@ -18,88 +18,232 @@ def initialize
     def add_first(value)
-      raise NotImplementedError
+      @head = Node.new(value, @head)
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      if @head.nil?
+        return false
+      end
+
+      current = @head
+
+      while current != nil
+        if current.data == value
+          return true
+        end
+        current = current.next
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_max
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+
+      max = @head.data
+      current = @head
+      while current != nil
+        if max < current.data
+          max = current.data
+        end
+
+        current = current.next
+      end
+
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+
+      min = @head.data
+      current = @head
+      while current != nil
+        if min > current.data
+          min = current.data
+        end
+
+        current = current.next
+      end
+
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      if @head.nil?
+        return 0
+      end
+
+      length = 0
+      current = @head
+      while current != nil
+        length += 1
+        current = current.next
+      end
+
+      return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)
     def get_at_index(index)
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+
+      count = 0
+      current = @head
+      while current != nil
+        if count == index
+          return current.data
+        end
+        count += 1
+        current = current.next
+      end
+
+      return nil
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit

```

Comment:
👍 , however the space complexity is O(1) because you're not creating a new list here.